controllers/datadogagent/controller_reconcile_v2.go

```diff
@@ -0,0 +1,161 @@
+			// Return and don't requeue
+			return result, nil
+		}
+		// Error reading the object - requeue the request.
+		return result, err
+	}
+
+	// check it the resource was properly decoded in v2
+	// if not it means it was a v1
+	if !apiequality.Semantic.DeepEqual(instance.Spec, datadoghqv2alpha1.DatadogAgentSpec{}) {
+		instanceV1 := &datadoghqv1alpha1.DatadogAgent{}
+		if err = r.client.Get(ctx, request.NamespacedName, instanceV1); err != nil {
+			if apierrors.IsNotFound(err) {
+				// Request object not found, could have been deleted after reconcile request.
+				// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+				// Return and don't requeue
+				return result, nil
+			}
+			// Error reading the object - requeue the request.
+			return result, err
+		}
+		if err = datadoghqv1alpha1.ConvertTo(instanceV1, instance); err != nil {
+			reqLogger.Error(err, "unable to convert to v2alpha1")
+			return result, err
+		}
+	}
+
+	if result, err = r.handleFinalizer(reqLogger, instance, r.finalizeDadV2); utils.ShouldReturn(result, err) {
+		return result, err
+	}
+
+	// TODO check if IsValideDatadogAgent function is needed for v2
+	/*
+		if err = datadoghqv2alpha1.IsValidDatadogAgent(&instance.Spec); err != nil {
+			reqLogger.V(1).Info("Invalid spec", "error", err)
+			return r.updateStatusIfNeeded(reqLogger, instance, &instance.Status, result, err)
+		}
+	*/
+
+	return r.reconcileInstanceV2(ctx, reqLogger, instance)
+}
+
+func (r *Reconciler) reconcileInstanceV2(ctx context.Context, logger logr.Logger, instance *datadoghqv2alpha1.DatadogAgent) (reconcile.Result, error) {
+	var result reconcile.Result
+
+	features, requiredComponents, err := feature.BuildFeatures(instance, reconcilerOptionsToFeatureOptions(&r.options, logger))
+	if err != nil {
+		return result, fmt.Errorf("unable to build features, err: %w", err)
+	}
+	logger.Info("requiredComponents status:", "agent", requiredComponents.Agent, "cluster-agent", requiredComponents.ClusterAgent, "cluster-check-runner", requiredComponents.ClusterCheckRunner)
+
+	// -----------------------
+	// Manage dependencies
+	// -----------------------
+	storeOptions := &dependencies.StoreOptions{
+		SupportCilium: r.options.SupportCilium,
+		Logger:        logger,
+	}
+	depsStore := dependencies.NewStore(storeOptions)
+	resourcesManager := feature.NewResourceManagers(depsStore)
+	var errs []error
+	for _, feat := range features {
+		if featErr := feat.ManageDependencies(resourcesManager); err != nil {
+			errs = append(errs, featErr)
+		}
+	}
+	// Now create/update dependencies
+	errs = append(errs, depsStore.Apply(ctx, r.client)...)
+	if len(errs) > 0 {
+		logger.V(2).Info("Dependencies apply error", "errs", errs)
+		return result, errors.NewAggregate(errs)
+	}
+	// -----------------------------
+	// Start reconcile Components
+	// -----------------------------
+
+	newStatus := instance.Status.DeepCopy()
+
+	if requiredComponents.ClusterAgent.IsEnabled() {
```

Comment:
Is a component move from `Enabled` to `Disabled`, will it be clean-up itself by `store.Cleanup`?

Reply:
No, we need to build the Component cleanup. not done yet :)